### PR TITLE
PS-7320 Doc Data Masking as GA in 8.0.21 and 5.7.32 (5.7)

### PR DESCRIPTION
**Pull request by patrickbirch**
modified:   source/index.rst
new file:   source/security/data-masking.rst

### CODE REVIEW

**Review by kathy-williamson — APPROVED** [review]

File: doc/source/security/data-masking.rst [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            mysql> SELECT mask_pan ('123456789012345');
+
+            +------------------------------------+
+            | mask_pan(gen_rnd_pan()).           |
+            +------------------------------------+
+            | XXXXXXXXXXX2345                    |
+            +------------------------------------+
+
+    * - mask_pan_relaxed(string)
+      - Returns the first six numbers and the last four    numbers. The rest of
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]:
```suggestion
      - Returns the first six numbers and the last four numbers. The rest of
```

File: doc/source/security/data-masking.rst [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+            mysql> SELECT mask_pan_relaxed(gen_rnd_pan());
+
+            +------------------------------------------+
+            | mask_pan_relaxed(gen_rnd_pan())          |
+            +------------------------------------------+
+            | 520754XXXXXX4848                         |
+            +------------------------------------------+
+    * - mask_ssn(string)
+      - Returns a  string with only the last four numbers visible. The rest
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]:
```suggestion
      - Returns a string with only the last four numbers visible. The rest
```

File: doc/source/security/data-masking.rst [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+              mysql> SELECT gen_range(10, 100);
+
+              +--------------------------------------+
+              | gen_range(10,100)                    |
+              +--------------------------------------+
+              | 56                                   |
+              +--------------------------------------+
+
+              mysql> SELECT gen_range(-100,-80);
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]: Are you sure this worked for negative numbers?

File: doc/source/security/data-masking.rst [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+be loaded from a file with the following characteristics:
+
+* Plain text
+
+* One term per line
+
+* Must contain at least one entry
+
+The following table displays the commands for using dictionaries to generate random terms:
+
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]: better to show example files with contents

File: doc/source/security/data-masking.rst [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+      - .. code-block:: MySQL
+
+             mysql> SELECT gen_dictionary_load('/usr/local/mysql/dict-files/testdict', 'testdict');
+
+             +------------------------------------------------------------------------+
+             | gen_dictionary_load('/usr/shared/mysql/dict-files/fndict', 'fndict')   |
+             +========================================================================+
+             | Dictionary load successfully                                           |
+             +------------------------------------------------------------------------+
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]: change this entire order of this section. It should be as if the user tries to use this functionality
1. creating dictionary files
2. gen_dictionary_load()
3. gen_dictionary()
4. gen_dictionary_blacklist()
5. gen_dictionary_drop().

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: doc/source/security/data-masking.rst (hunk 1 above)
Comment (patrickbirch) [reply doc/source/security/data-masking.rst]: Fixed.

File: doc/source/security/data-masking.rst (hunk 2 above)
Comment (patrickbirch) [reply doc/source/security/data-masking.rst]: Fixed.

File: doc/source/security/data-masking.rst (hunk 3 above)
Comment (patrickbirch) [reply doc/source/security/data-masking.rst]: PS 7309 says that it does.

File: doc/source/security/data-masking.rst (hunk 4 above)
Comment (patrickbirch) [reply doc/source/security/data-masking.rst]: Added example

File: doc/source/security/data-masking.rst (hunk 5 above)
Comment (patrickbirch) [reply doc/source/security/data-masking.rst]: Restructured.

File: doc/source/security/data-masking.rst (hunk 3 above)
Comment (satya-bodapati) [reply doc/source/security/data-masking.rst]: @patrickbirch Right, but that was fixed only in 8.0.
IMO, we should push this PR only to 8.0.

5.7, we should backport the bugfixes and then backport this documentation ticket. What do u say?

File: doc/source/security/data-masking.rst [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+* Generation of random data, such as an email address
+
+* De-identify data by transforming the data to hide content
+
+.. rubric:: Installing the plugin
+
+The following command installs the plugin:
+
+.. code-block:: bash
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]: s/bash/mysql?

File: doc/source/security/data-masking.rst [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+            mysql> SELECT gen_blacklist('apple', 'fruit', 'nut');
+
+            +-----------------------------------------+
+            | gen_blacklist('apple', 'fruit', 'nut')  |
+            +-----------------------------------------+
+            | walnut                                  |
+            +-----------------------------------------+
+
+----------------------------+
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]: this line should be somewhere else?

File: doc/source/security/data-masking.rst [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+
+      - .. code-block:: MySQL
+
+            mysql> SELECT gen_dictionary('trees');
+
+            +--------------------------------------------------+
+            | gen_dictionary('trees')                          |
+            +--------------------------------------------------+
+            | Norway spruce                                    |
+            +----------------------
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]: is the line formatting complete?

File: doc/source/security/data-masking.rst [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+      - .. code-block:: MySQL
+
+            mysql> SELECT gen_blacklist('apple', 'fruit', 'nut');
+
+            +-----------------------------------------+
+            | gen_blacklist('apple', 'fruit', 'nut')  |
+            +-----------------------------------------+
+            | walnut                                  |
+            +-----------------------------------------+
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst:296]: Better to use above dictionary example with trees

File: doc/source/security/data-masking.rst [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+Data masking returns a value from a range. To use a predefined file as the range to select a string value, load and use a dictionary. A dictionary supports only strings and is loaded from a file with the following characteristics:
+
+* Plain text
+
+* One term per line
+
+* Must contain at least one entry
+
+An example of a dictionary is a list of trees:
```
Comment (satya-bodapati) [inline doc/source/security/data-masking.rst]: list of trees stored in a file. In our example, it is /usr/local/mysql/dict-files/testdict.

Can you please rephrase?

File: doc/source/security/data-masking.rst (hunk 1 above)
Comment (satya-bodapati) [reply doc/source/security/data-masking.rst]: I still see multiple spaces between four and numbers..

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: doc/source/security/data-masking.rst (hunk 1 above)
Comment (patrickbirch) [reply doc/source/security/data-masking.rst]: Odd. Not in raw doc.

Comment (patrickbirch) [reply doc/source/security/data-masking.rst]: fixed.

Comment (satya-bodapati) [reply doc/source/security/data-masking.rst]: I still see the same :) May be you can use commit suggestion.

Comment (satya-bodapati) [reply doc/source/security/data-masking.rst]: Ok, I see it fixed now..

**Review by satya-bodapati — APPROVED** [review]
Ok to merge after fixing the minor comment (bash/mysql codeblock)

**Pull request merged**